libraries/AP_Camera/AP_RunCam.h

```diff
@@ -0,0 +1,319 @@
+    enum RCCameraCommands
+    {
+        RCDEVICE_PROTOCOL_COMMAND_GET_DEVICE_INFO = 0x00,
+        RCDEVICE_PROTOCOL_COMMAND_CAMERA_CONTROL = 0x01,
+        RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_PRESS = 0x02,
+        RCDEVICE_PROTOCOL_COMMAND_5KEY_SIMULATION_RELEASE = 0x03,
+        RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION = 0x04
+    };
+
+    // operation of 5 Key OSD cable simulation
+    enum RC5KeySimulationOperation
+    {
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_NONE = 0x00,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_SET = 0x01,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_LEFT = 0x02,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_RIGHT = 0x03,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_UP = 0x04,
+        RCDEVICE_PROTOCOL_5KEY_SIMULATION_DOWN = 0x05
+    };
+
+    // operation of RCDEVICE_PROTOCOL_COMMAND_5KEY_CONNECTION
+    enum RC5KeyConnectionEvent
+    {
+        RCDEVICE_PROTOCOL_5KEY_CONNECTION_OPEN = 0x01,
+        RCDEVICE_PROTOCOL_5KEY_CONNECTION_CLOSE = 0x02
+    };
+
+    // main simulation key press events
+    enum RCCamSimulationKeyEvent
+    {
+        RCDEVICE_CAM_KEY_NONE,
+        RCDEVICE_CAM_KEY_ENTER,
+        RCDEVICE_CAM_KEY_LEFT,
+        RCDEVICE_CAM_KEY_UP,
+        RCDEVICE_CAM_KEY_RIGHT,
+        RCDEVICE_CAM_KEY_DOWN,
+        RCDEVICE_CAM_KEY_CONNECTION_CLOSE,
+        RCDEVICE_CAM_KEY_CONNECTION_OPEN,
+        RCDEVICE_CAM_KEY_RELEASE,
+    };
+
+    // protocol versions, only version 1.0 is supported
+    enum RCProtocolVersion
+    {
+        RCDEVICE_PROTOCOL_RCSPLIT_VERSION = 0x00, // unsupported firmware version <= 1.1.0
+        RCDEVICE_PROTOCOL_VERSION_1_0 = 0x01,
+        RCDEVICE_PROTOCOL_UNKNOWN
+    };
+
+    // status of command response
+    enum RCResponseStatus
+    {
+        RCDEVICE_RESP_SUCCESS = 0,
+        RCDEVICE_RESP_INCORRECT_CRC = 1,
+        RCDEVICE_RESP_TIMEOUT = 2
+    };
+
+    // initialize the RunCam driver
+    void init();
+    // camera button simulation
+    bool simulate_camera_button(RCCameraControlOperation operation);
+    // start the video
+    void start_recording();
+    // stop the video
+    void stop_recording();
+    // update loop
+    void update();
+    // OSD update loop
+    void update_osd();
+
+    static const struct AP_Param::GroupInfo var_info[];
+
+private:
+    // supported features, usually probed from the device
+    AP_Int16 _features;
+    // number of initialization attempts
+    AP_Int8 _init_attempts;
+    // delay between initialization attempts
+    AP_Int32 _init_attempt_interval_ms;
+    // delay time to make sure the camera is fully booted
+    AP_Int32 _boot_delay_ms;
+    // delay time to make sure a button press has been activated
+    AP_Int32 _button_delay_ms;
+    // detected or configured protocol version
+    AP_Int8 _protocol_version;
+
+    // uart for the device
+    AP_HAL::UARTDriver *uart;
+    // shared scratch space for communication
+    uint8_t _buffer[RCDEVICE_PROTOCOL_MAX_PACKET_SIZE];
```

Comment:
I don't think we need a shared scratch space.  We can afford 64 bytes on the stack - and the one place I really looked at required 4 IIUC!